### DISCUSSION TITLE
TuyaMCU for ESP32

### DISCUSSION THREAD

**Discussion by elimisback** [arendst/Tasmota#14086] [Q&A] [answered]
I was wondering if there was anyone looking at porting TUYA MCU to the ESP32 variant. I have some treatlife switches that I was trying to do a chip swap with. The esp32-c3-12f adds features and works better from a hardware prospective based on the pins it needs pulled up and down vs the esp12F

**Comment by sfromis** [comment]
If your issue with ESP-12F is the strapping pins, ESP-12S is easier to deal with.

**Comment by barbudor** [comment, marked as answer]
Why "porting"?
Tuya MCU is support by Tasmota including tasmota32.

**Reply by arendst** [reply]
Currently USE_TUYA_MCU and USE_TUYA_DIMMER are disabled for ESP32 (See ``tasmota_globals``). 

They may be enabled but some features need to be checked first (like ESP8266 TasmotaGlobal.module_type clashing with ESP32 module types).

**Reply by sfromis** [reply]
I'm guessing it is the same as Mikein3D who on Discord was asking about the same, and got build errors trying to enable TuyaMCU support, removing the #undef below a comment about features not ported.

**Reply by elimisback** [reply]
Yeah i am same person from discord. I figured it was ok to  ask here too. i looked at the 12s. I just figured the esp32 is newer.

**Reply by sfromis** [reply]
Well, the audience for TuyaMCU with ESP32 is likely to stay very, very, very low ;-) Of course, if you work out the kinks, and submit a PR .....

**Reply by elimisback** [reply]
I only wish I was able to code like that, but we will see.

**Reply by barbudor** [reply]
@elimisback 
If you have an ESP32-C3-12F ready for testing, the attached binary includes TUYA_MCU
Note that you will need to manually adapt templates
Especially the BASE should be 2 and no more 54
If this looks to work fine, I can then have a look with Theo on how to "translate" templates automatically.
_outdated binary removed_

For reference, WIP branch : https://github.com/barbudor/Tasmota/tree/tuya_esp32

**Reply by elimisback** [reply]
Dont have to much time to tinker today, but i was able to get the chip soldered on. I had to update the gpio assignments to gpio21 Tuya Tx and gpio20 Tuya RX. I also set option 97 to 1. From there i was able to see the heartbeat and adjust dimmer values dont have time to wire into a switch as i need to get ready for a wedding but will do that later thank you so much

**Reply by elimisback** [reply]
I have it all wired up and communicating with home assistant thru tasmota integration no issues of note.

**Reply by barbudor** [reply]
Ok, I'll send a PR and let's see what Theo's thinking about it
https://github.com/arendst/Tasmota/pull/14106

**Reply by barbudor** [reply]
@elimisback Theo has ddecided to implement slighly differently. There is finally no need for selecting module "Tuya (2)".

It is now available for all Tasmota32 versions in dev : http://ota.tasmota.com/tasmota32/
Could you please upgrade to this official version and let us know how it goes ?
Thanks

**Reply by elimisback** [reply]
working on it right now

**Reply by elimisback** [reply]
No issues flashing with most recent development version.

**Reply by jcauble** [reply]
@elimisback I am currently working through this exact issue and wondering how you got it to work?  I transplanted an esp32c3 onto a tuya dimmer switch and the chip boots up and all the buttons work, but I don't have the communication with tuyamcu.  I am running the latest tasmota32c3.bin (shows Tasmota 10.1.0.1) and also tried the tasmota32c3-tuyatest.bin listed here.  I don't see an option on the module configuration for either tuyatx or tuyarx.

**Reply by barbudor** [reply]
Please use development version 2022.01.** from http://ota.tasmota.com/tasmota32/
TuyaMCU for ESP32 has been implemented after v10.0.1
